### ISSUE TITLE
Is it runnable without web workers?

### ISSUE THREAD

**Issue by iliakan** [mame/xterm-pty#5]
Because of other integration issues, my Emscripten-compiled code must run in the main browser window.

Can I use xterm-pty? 

You say that there's some problem with `SharedArrayBuffer`, you also say that `SharedArrayBuffer` is available in the main browser window (assuming that it's secure). 

Tried to read your explanation a few times, but still don't getting, what's the issue, and can it be worked around?

**Comment by mame** [comment]
As far as I know, you need a Web Worker. This is because the JavaScript code generated by Emscripten reads input from stdin synchronously. Unfortunately, it is not possible to wait for input synchronously in the main thread.

It would be possible if there is an option in Emscripten to output JavaScript code that waits asynchronously for input. A year ago when I created xterm-pty, I couldn't find such an option.

**Comment by iliakan** [comment]
Yes, there's such an option, but...

Interesting, how did you make it work in with workers then? Which feature is absent in the main window, but works in the worker?

**Comment by mame** [comment]
Web Worker can wait synchronously without blocking the user interface. The main thread cannot.

**Comment by iliakan** [comment]
Got it! So it's `Atomics.wait`, thank you for expanations ;)

**Comment by RReverser** [comment]
It should be possible to at least make this library with PROXY_TO_PTHREAD. Unlike manual workers, PROXY_TO_PTHREAD is well integrated with Emscripten and automatically proxies any relevant calls to the main thread, and still allows to execute blocking calls without blocking the browser UI.

**Comment by mame** [comment]
Hmmm, I guess I'm not familiar enough with Emscripten yet. If you could show a better integration as a PR, I'd be happy to review it.

**Comment by RReverser** [comment]
All that said, proxy-to-pthread still wouldn't work for scenarios where Emscripten app expects to have access to DOM via JS. Asyncify mentioned above would be a more universal even if more costly in terms of size.

I'll play a bit with the latter as I'm pretty familiar with it.

**Comment by mame** [comment]
I read the code of Emscripten little based on the information you provided.

We need to wait synchronously in [TTY.stream_ops.read](https://github.com/emscripten-core/emscripten/blob/01c4612a32638ec263323b04a20c43ddb62c6fe1/src/library_tty.js#L62). This function is called via [__wasi_fd_read](https://github.com/emscripten-core/emscripten/blob/01c4612a32638ec263323b04a20c43ddb62c6fe1/src/library_wasi.js#L330) and [FS.read](https://github.com/emscripten-core/emscripten/blob/01c4612a32638ec263323b04a20c43ddb62c6fe1/src/library_fs.js#L1159).

I think we need to prevent `fd_read` from proxying and run it in a worker thread. However, looks like this function need to be proxied for the file system access according to [the document](https://emscripten.org/docs/porting/pthreads.html#proxying):

> In addition, Emscripten currently has a simple model of file I/O only happening on the main application thread (as we support JS plugin filesystems, which cannot share memory); this is another set of operations that are proxied.

I have two questions.

* `fs_read` is defined in library_wasi.js. Can I specify `fs_read__proxy` to it from an external project? Maybe is it possible to `addToLibrary({ fs_read__proxy: 'sync' })` by external `--js-library=library_xterm_pty.js`?
* I think `fs_read` need to be proxied when it attempts to access the file system, and not proxied when it attempts to access TTY. Do you know whether and how such a conditional proxy is possible?

Thanks!

**Comment by RReverser** [comment]
Yeah I came to similar conclusion - it's already proxied, or `prompt` wouldn't even work in a pthread.

> * `fs_read` is defined in library_wasi.js. Can I specify `fs_read__proxy` to it from an external project? Maybe is it possible to `addToLibrary({ fs_read__proxy: 'sync' })` by external `--js-library=library_xterm_pty.js`?

You can, except you don't want `sync` - that's what enables synchronous proxying. If you want to disable proxying, you want to unset that variable, with something like:

```js
Object.assign(LibraryManager.library, {
	fd_read__proxy: undefined,
});
```

That said, as mentioned above, I'm currently looking into using Asyncify instead. It should cover more usecases and, aside from one bug I'll need to fix in Emscripten, should integrate pretty easily.

It will be even easier with WASMFS which has a dedicated `_wasmfs_stdin_get_char` import that can be augmented directly instead of dealing with higher-level `fd_read`, but from looking at Emscripten source, looks like WASMFS doesn't support TTY yet so probably not compatible with this library.

**Comment by RReverser** [comment]
I'll continue working on it next week, but got a simple PoC of reading chars asynchronously working; next up - integrate with this library.

**Comment by RReverser** [comment]
Got it to work on the same thread with Asyncify. It needs latest Emscripten (as I made some changes upstream) + some cleanup before I can send a PR, but looks promising:

![image](https://github.com/mame/xterm-pty/assets/557590/e5425e55-fc70-48cc-9d97-c516826a2653)

This is all the code user needs to run the demo (only to pass `pty` as `Module['pty']`), in addition to `--js-library` to include a small library I made for the integration:

```js
import { openpty } from './bundle.js';
import initEmscripten from './temp.js';

var xterm = new Terminal();
xterm.open(document.getElementById('terminal'));

// Create master/slave objects
const { master, slave } = openpty();

// Connect the master object to xterm.js
xterm.loadAddon(master);

await initEmscripten({
  pty: slave
});
```

**Comment by RReverser** [comment]
UPD: simple examples work fine, but ncurses-based (`sl` and `vim`) are more difficult to get right, so still working on it.

**Comment by mame** [comment]
@RReverser Thank you very much for your great work. It looks awesome that it works without WebWorker! It does not require tedious SharedArrayBuffer, right?

It would be definitly great if we can resolve this issue. So I reopen this.

**Comment by RReverser** [comment]
> t does not require tedious SharedArrayBuffer, right?

In my current solution it's somewhere in-between: it doesn't require SAB for demos where you read after writing, but still requires it for continuous writing like the Sloane demo.

Basically I'm not implementing program pauses for writes themselves to keep overhead for typical scenarios low, and instead leaving the pausing strategy to the user. This is similar to how user needs to pause their program when they're drawing to canvas or operating the DOM manually in Emscripten, and user can choose between Emscripten's event loop (https://emscripten.org/docs/porting/emscripten-runtime-environment.html), manual Asyncify pauses like `emscripten_sleep(0)` or `-s PROXY_TO_PTHREAD` - which is what I used for Sloane demo, and which makes it use a separate thread like before but without tedious manual wrappers, just one more compile flag.

I might change this pausing / flushing strategy later, but for now this seems to strike a good balance.

FWIW I also got Vim to work (without SAB / extra threads, just on the main thread), but for some reason SL demo is still [the only one] misbehaving so still investigating.

**Comment by RReverser** [comment]
The implementation now more complicated to correctly handle polling in different modes (both Asyncify and PROXY_TO_PTHREAD), but all demos work again now.

Cleaning up code and should be able to submit soon-ish.
